src/libraries/System.Linq/src/System/Linq/Range.SpeedOpt.cs

```diff
@@ -16,23 +19,52 @@ public override IEnumerable<TResult> Select<TResult>(Func<int, TResult> selector
 
             public int[] ToArray()
             {
-                int[] array = new int[_end - _start];
-                Fill(array, _start);
+                int start = _start;
+                int[] array = new int[_end - start];
+                Fill(array, start);
                 return array;
             }
 
             public List<int> ToList()
             {
-                List<int> list = new List<int>(_end - _start);
-                Fill(SetCountAndGetSpan(list, _end - _start), _start);
+                int start = _start;
+                int end = _end;
+                List<int> list = new List<int>(end - start);
+                Fill(SetCountAndGetSpan(list, end - start), start);
                 return list;
             }
 
-            private static void Fill(Span<int> destination, int value)
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private static void Fill(Span<int> destination, int start)
             {
-                for (int i = 0; i < destination.Length; i++, value++)
+                // The improvements in the compiler now allow us to write the loop as Vector256.
+                // Manually unrolling to Vector128x2 for SSE2/NEON is no longer profitable.
+                if (destination.Length < Vector256<int>.Count)
                 {
-                    destination[i] = value;
+                    goto Scalar;
+                }
+
+                int stride = Vector256<int>.Count;
+
+                Vector256<int> init = Vector256.Create(
+                    (ReadOnlySpan<int>)new int[] { 0, 1, 2, 3, 4, 5, 6, 7 });
```

Comment:
```suggestion
                Vector256<int> init = Vector256.Create(0, 1, 2, 3, 4, 5, 6, 7);
```
The a [constant](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAJI8ovLrl5hcAbho1iAZiakGAYQYBvGg11NlANRhgM0UgFYkAHnEYAfAwCy5ABQBKLTr3ejJs5ZseB3FeDAYAXgZfUygLJBY1WGwMGBcXNhhsABMAeS4AGwBPAGUAB2wuQPs3LhgAdwZbAG0AXS0GOjQGci7SLsUu9AZzLqQugHYGAF83OWpvb2JJkIw57ymvXU39KOMYuKqHR1J3T3mF3Wj/a1tgrlCI3b9YywSklJdO7t7+wa6RhhjBjjWbbRbLe6rbYbahTIA===) will be emitted and used.

Reply:
Nice, it seems it wasn't the case in the past but now the codegen is optimal in either case (in Tier 1). I haven't looked at Native AOT though. Fixed.